src/dialogs/create-file.tsx

```diff
@@ -0,0 +1,233 @@
+    });
+}
+
+async function create_file(filename: string, content?: string, owner?: string) {
+    // let tag = await read_tag(filename);
+    const file = cockpit.file(filename, { superuser: "try" });
+    try {
+        await file.replace("");
+    } catch (err) {
+        console.warn("Cannot create new file", err);
+        throw err;
+    }
+
+    if (owner !== undefined) {
+        try {
+            await cockpit.spawn(["chown", owner, filename], { superuser: "require" });
+        } catch (err) {
+            console.warn("Cannot chown new file", err);
+            throw err;
+        }
+    }
+
+    // re-read the tag as it contains the ownership bits
+    const tag = await read_tag(filename);
+
+    if (content) {
+        // We need to obtain the tag to retain file ownership
+        // Can't do this async because we can't get the tag via await
+        try {
+            await cockpit.file(filename, { superuser: "try" }).replace(content, tag);
+        } catch (err) {
+            console.warn("Cannot set initial file text", err);
+            throw err;
+        }
+    }
+}
+
+export const CreateFileModal = ({ dialogResult, path } : {
+    dialogResult: DialogResult<string>,
+    path: string
+}) => {
+    const [user, setUser] = React.useState<cockpit.UserInfo| undefined>();
+    const [filename, setFilename] = React.useState<string>("");
+    const [owner, setOwner] = React.useState<string | undefined>();
+    const [initial_text, setInitialText] = React.useState<string>("");
+    const [filename_error, setFileNameError] = React.useState<string | null>(null);
+    const [create_error, setCreateError] = React.useState<string | null>(null);
+
+    const { cwdInfo } = useFilesContext();
+
+    useInit(() => {
+        cockpit.user().then(user => setUser(user));
+    });
+
+    const handleEscape = (event: KeyboardEvent) => {
+        if (filename !== "" || initial_text !== "") {
+            event.preventDefault();
+        } else {
+            dialogResult.resolve("");
+        }
+    };
+
+    const handleSave = async () => {
+        cockpit.assert(filename, "filename undefined");
+        const filename_valid = checkFilename(filename, cwdInfo?.entries || {}, undefined);
+        if (filename_valid !== null) {
+            setFileNameError(filename_valid);
+            return;
+        }
+
+        const full_path = path + filename;
+        try {
+            await create_file(full_path, initial_text, owner);
+        } catch (err) {
+            const exc = err as cockpit.BasicError; // HACK: You can't easily type an error in typescript
+            setCreateError(exc.message);
+            return;
+        }
+
+        dialogResult.resolve(full_path);
+    };
+
+    const candidates = [];
+    if (superuser.allowed && user && cwdInfo) {
+        candidates.push(...get_owner_candidates(user, cwdInfo));
+        if (owner === undefined) {
+            setOwner(candidates[0]);
+        }
+    }
```

Comment:
That whole owner initialization should better move into `useInit()`. It's brittle and inefficient to do that in every render (and rendering shouldn't change state either), and you have to do that ugly `=== undefined` check here.